Adding architecture docs. (#22371)
Azure/azure-sdk-for-java: sdk/core/azure-core-amqp/docs/architecture.md

# Architecture Docs

## Qpid Proton-J Integration

Qpid Proton-J publishes events and messages via its event-driven process called [`Reactor`][Reactor]. `azure-core-amqp`

> hooks into Qpid Proton-J Reactor via `Handlers`.  [`BaseHandler`][BaseHandler] contains all of the events that can be

Reviewer: Is it the case that once we create a `Reactor` object then that object creates and owns a `BaseHandler` object, and we can register our own more granular user-handler objects? Like one for connection-events, another for session-events etc.. Where base-handler demultiplex events to these specialized user-handlers?

listened to.  `Handlers` can be associated with classes implementing the `Extendable` interface via

> `BaseHandler.setHandler(Extendable, Handler)`.  Proton-J `Connection`, `Session`, `Link`, `Sender` and `Receiver` are

Reviewer: so `Sender` and `Receiver` are not really each endpoint of  `Link` but are two specializations of `Link` type, right?

all `Extendable`.

> The UML diagram below shows this relationship; the interfaces shown in green are Proton-J classes. Each Proton-J

Reviewer: nit; Should we say "the interfaces shown in green are Proton-J **types**"?

Reviewer 2: Let's avoid referring to diagram elements by color to accommodate folks with color blindness. Instead of "the interfaces shown in green...", I recommend something like "the boxes annotated with `<<Interface>>`".

> instance (ie. `Connection`, `Session`, `Sender`, `Receiver`) is associated with one corresponding `*Handler`.  Each

Reviewer: will the `Reactor` object also own the `Connection` object whichh can be specicifed when registering user-handler to handle events from that connection object? I mean my main question is - who is responsible to own and manage the connection? user or reactor.

> instance (ie. `Connection`, `Session`, `Sender`, `Receiver`) is associated with one corresponding `*Handler`.  Each

Reviewer 2: ```suggestion
instance (e.g., `Connection`, `Session`, `Sender`, `Receiver`) is associated with one corresponding `*Handler`.  Each
```

> azure-core-amqp `AmqpConnection` is associated with one [`Reactor`][Reactor].  When that instance closes, the AMQP

Reviewer 2: ```suggestion
`azure-core-amqp` `AmqpConnection` is associated with one [`Reactor`][Reactor]. When that instance closes, the AMQP
```

connection is also closed.

Each [ReactorConnection][ReactorConnection] has one Proton-J [`Reactor`][Reactor] instance.  Each [`Reactor`][Reactor]

> has one [`ReactorDispatcher`][ReactorDispatcher] and one [`ReactorExecutor`][ReactorExecutor].

Reviewer: Can we have a small description of how a `Reactor` object uses the dispatcher object and executor object? (for details we can always refer the Javadoc and code)

![azure-core-amqp integration with Proton-J][AzureCoreAmpqArchitecture]

> ## Prefetch and AMQP Link Credits

Reviewer 3: This looks great.  One thing to call out is that It would be nice if terms like link credits were defined before their use, similar to what's in : https://github.com/Azure/amqpnetlite/blob/master/docs/articles/building_application.md

> In Project Reactor, prefetch is the initial number of items to request upstream, afterwards, 75% of the initial prefetch

Reviewer 2: ```suggestion
In Project Reactor, prefetch is the initial number of items to request upstream. Afterwards, 75% of the initial prefetch
```

is used for subsequent `request(long)`.

In Event Hubs, prefetch is the number of AMQP link credits to put on the link when it is first created.  After those
initial link credits have been consumed, we have different ways of calculating how many credits are added to the link.

The diagram below illustrates how it happens. Things to note:

> * Large `EventData` use multiple AMQP link credits.

Reviewer: Is it because of the large message split across multiple `Transfer` frames? (where each frame consumes one credit)

> * There is no backpressure for [`EventHubConsumerAsyncClient.receiveFromPartition()`][EventHubConsumerAsyncClient].

Reviewer: By no-back-pressure do we mean that we manage the credit underneath hence SDK (and protonj) decide when to send the `Flow` frame, not the subscriber subscribing to the Flux returned from receive, right?

Reviewer: Is there any reactor operator chain we use that results in in-memory queuing of the message? (that give the feeling of  simulated (but local) backpressure support for the user)

* [`EventHubConsumerAsyncClient.receiveFromPartition()`] returns `EventData` on `Scheduler.single("<name>")`.
  * Since events are not published on another Scheduler, they flow downstream using the Scheduler that
    [`ReceiveLinkHandler.onDelivery`][ReceiveLinkHandler] executed on.

>   * All Proton-J events run on the single scheduler because it is not thread-safe.

Reviewer: Is this scheduler somehow connected to the `Reactor` dispatcher|executor mentioned earlier in the doc?

* [`EventProcessorClient`][EventProcessorClient] uses back-pressure due to `concatMap` and `publishOn` within its
  [PartitionPumpManager.startPartitionPump][PartitionPumpManager].

> 

Reviewer: in the below nice digram, when we say "Drain queue of event data", does it mean our SDK drains the queue by poping one event data at a time (and delivering EACH event-data to the application)?
